### PR TITLE
Fixes #20513: Sometimes the webapp fails to generate the reporting file of new techniques

### PR DESCRIPTION
**Pull request by amousset**
https://issues.rudder.io/issues/20513

### CODE REVIEW

**Comment by amousset** [comment]
This does not work

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala [hunk 1]
```diff
@@ -181,12 +181,40 @@ class RudderCRunner(
...
+         |}"""
+
+    val reportingFile = File(outputPath) / "rudder_reporting.cf"
+
+
+    TechniqueWriterLoggerPure.error(s"PLOUF target '${technique.name}' in path ${reportingFile.path.toString}")
+
+    IOResult.effect(s"Could not write empty reporting Technique file '${technique.name}' in path ${reportingFile.path.toString}") {
+      reportingFile.createFileIfNotExists(true).write(content.stripMargin('|'))
+      Seq(File(outputPath).relativize(reportingFile.path).toString)
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala:201]: this is not used, can be removed

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala [hunk 2]
```diff
@@ -181,12 +181,40 @@ class RudderCRunner(
...
+    val bundleParams = if (technique.parameters.nonEmpty) technique.parameters.map(_.name.canonify).mkString("(",",",")") else ""
+    val content =
+      s"""bundle agent ${technique.bundleName.value}_rudder_reporting${bundleParams}
+         |{
+         |}"""
+
+    val reportingFile = File(outputPath) / "rudder_reporting.cf"
+
+
+    TechniqueWriterLoggerPure.error(s"PLOUF target '${technique.name}' in path ${reportingFile.path.toString}")
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala:197]: as that instuction is a pure value, it won't product side effect. If you want (you want) to include it in tha chain of instruction to be interpreted by the `IO` interpreter, you need to bundle it with the following one (which is returned) by adding a `*>` (which means "and pack it to next instruction, I don't care of its returned value"). 
You can also do: 
```
for {
  _ <- TechniqueWriterLoggerPure.error(s"PLOUF target '${technique.name}' in path ${reportingFile.path.toString}")
 _ <- IOResult.effect(s"Could....) { ... }
} yield () 
``` 
which would allows to use intermediate returned values if you needed them.

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala [hunk 3]
```diff
@@ -181,12 +181,40 @@ class RudderCRunner(
...
+
+    IOResult.effect(s"Could not write empty reporting Technique file '${technique.name}' in path ${reportingFile.path.toString}") {
+      reportingFile.createFileIfNotExists(true).write(content.stripMargin('|'))
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala:200]: by default, `stripMargin` uses `|` , you can remove it (but it's ok to keep it too)

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Review by VinceMacBuche — APPROVED** [review]

**Comment by amousset** [comment]
OK, squash merging this PR

**Pull request merged**